ir_datasets/datasets/sara.py

```diff
@@ -0,0 +1,225 @@
+    cats_globs = glob.glob(directory + '/*/*/*.cats')
+    matches = []
+    for g in cats_globs:
+        with open(g) as file:
+            for line in file:
+                line = line.split()[0]
+                line = line.split(",")
+                if int(line[0]) == top and int(line[1]) == second:
+                    dir = os.path.basename(g)
+                    filename = os.path.splitext(dir)[0]
+                    matches.append(filename)
+    return matches
+
+def print_email_from_filename(directory,filename):
+    full_filename = glob.glob(directory +'/*/*/*' + str(filename) + '.txt')
+    with open(full_filename[0]) as file:
+        for line in file:
+            print(line)
+
+def save_email_from_filename(filename, directory):
+    full_filename = glob.glob(directory +'/*/*/' + str(filename) + '.txt')
+    email_contents = ""
+    with open(full_filename[0]) as file:
+        for line in file:
+            email_contents = email_contents + line
+    return email_contents
+
+def save_all_from_cat(primary, secondary):
+    filenames = filenames_from_cat(primary, secondary)
+    contents = ""
+    for file in filenames:
+        contents = contents + save_email_from_filename(file)
+        contents = contents + "\n" + ("*" * 100) + "\n\n"
+    return contents
+
+def list_all_filenames(directory):
+    globs = glob.glob(directory +'/*/*/*.cats')
+    filenames = []
+    for file in globs:
+        filenames.append(os.path.splitext(os.path.basename(file))[0])
+    return filenames
+
+def length_of_email_from_filename(filename, directory):
+    email_text = save_email_from_filename(filename)
+    tokenised = word_tokenize(email_text)
+    return len(tokenised)
+
+def number_of_recipients(filename):
+    addresses = []
+    contents = save_email_from_filename(filename)
+    contents = contents.split("\n")
+    for line in contents:
+        if line.startswith("To:"):
+            addresses.append(re.findall(r"[a-z0-9\.\-+_]+@[a-z0-9\.\-+_]+\.[a-z]+", line))
+    
+    return len([j for sub in addresses for j in sub])
+
+def get_sender_from_filename(filename):
+    addresses = []
+    contents = save_email_from_filename(filename)
+    contents = contents.split("\n")
+    for line in contents:
+        if line.startswith("From:"):
+            addresses.append(re.findall(r"[a-z0-9\.\-+_]+@[a-z0-9\.\-+_]+\.[a-z]+", line))
+    
+    return [j for sub in addresses for j in sub][0]
+
+def search_by_id(id):
+    all_emails = list_all_filenames()
+    for email in all_emails:
+        text = save_email_from_filename(email)
+        if id in text:
+            print(text)
+
+def sensitivity_stats(filenames):
+  non_sensitive_count = 0
+  sensitive_count = 0
+  
+  for f in filenames:
+    if f in non_sensitive_filenames:
+      non_sensitive_count += 1
+    if f in sensitive_filenames:
+      sensitive_count += 1
+
+  print(f"Non-Sensitive: {non_sensitive_count} | Sensitive: {sensitive_count} ")
+
+def get_docs_frame(directory):
```

Comment:
Can you re-work this function as an iterator? I.e., using the `yield` keyword? I think it can replace the `data.append` statements. This should avoid loading the entire dataset into memory at once.